### PR TITLE
bug fix: "DSP panic" when ALSA XRUN recovery

### PR DESCRIPTION
**Pull request by zhigang-wu**
I did not test it with latest version, because of the some known issues or regression issues on the latest version.

### CODE REVIEW

File: src/ipc/handler.c [hunk 1]
```diff
@@ -449,6 +449,18 @@ static int ipc_stream_trigger(uint32_t header)
...
 	}
 
+	/* when stop command comes, the pipeline has to be prepared.
+	 * This case has to be taken into account:
+	 * when ALSA has XRUN, it will send stop/start commands to FW
+	 * and try to recover the pipepine. If we did not prepare
+	 * the pipeline before start, the recover process will fail.
+	 */
+	if (cmd == COMP_TRIGGER_STOP) {
+		ret = pipeline_prepare(pcm_dev->cd->pipeline, pcm_dev->cd);
```
Comment (plbossart) [inline src/ipc/handler.c]: that looks quite questionable. it's rather when you start that you should make sure things are prepared, no?

File: src/audio/host.c [hunk 2]
```diff
@@ -740,7 +740,7 @@ static int host_stop(struct comp_dev *dev)
 	/* reset elements, to let next start from original one */
 	host_elements_reset(dev);
 
-	dev->state = COMP_STATE_PAUSED;
+	dev->state = COMP_STATE_PREPARE;
```
Comment (plbossart) [inline src/audio/host.c]: this is really odd. You will end-up with different states in different components, see e.g. the SSP dai code where the component are set to PAUSED.
Does this even work?

File: src/audio/component.c [hunk 3]
```diff
@@ -142,7 +142,8 @@ int comp_set_state(struct comp_dev *dev, int cmd)
 		break;
 	case COMP_TRIGGER_STOP:
 	case COMP_TRIGGER_XRUN:
-		if (dev->state == COMP_STATE_ACTIVE) {
+		if (dev->state == COMP_STATE_ACTIVE ||
+		    dev->state == COMP_STATE_PREPARE) {
```
Comment (plbossart) [inline src/audio/component.c]: WTH?
If you are in PREPARE state you set the component to PREPARE? I call this a no-op...

File: src/ipc/handler.c (hunk 1 above)
Comment (zhigang-wu) [reply src/ipc/handler.c]: It takes effect similar like the pipeline XRUN recover process.
you can refer to the pipeline_xrun_recover() to understand the logic.
without prepare, the firmware will hit XRUN again when start command comes.

File: src/audio/host.c (hunk 2 above)
Comment (zhigang-wu) [reply src/audio/host.c]: I think when the component is truly paused, it should be paused state.
but if the command is stop. we have to set it prepare state to ready for the next start comes.
that makes sense.

File: src/audio/component.c (hunk 3 above)
Comment (zhigang-wu) [reply src/audio/component.c]: I will think about your question.
Actually it works! I tested it on my GP, UP^2, BYT.

**Comment by zhigang-wu** [comment]
This RFCs idea comes from the pipeline_xrun_recover() process.
I checked the commands sent from host. it sends the different commands between 
normal start and ALSA xrun's recover start.
that is the reason I did it like this.
I did this test on BYT, GP, UP^2.  No negative effects found.
If you think it is incorrect in some logic, please point it out. I will be glad to change it.
This DSP panic exists about more than four months.
I think this RFCs can give us the correct direction to fix this kind of issue.

File: src/ipc/handler.c (hunk 1 above)
Comment (zhigang-wu) [reply src/ipc/handler.c]: I make a stupid typo error, the Line459 should be:
ret = pipeline_prepare(pcm_dev->pipeline, pcm_dev->pipeline->source_comp);
I will update it after back office.

**Comment by plbossart** [comment]
@zhigang-wu: The flow in open source is that you have to convince people that your solution is correct, not expect maintainers to tell you what is wrong in the logic. The code you provided does not convince me, so the ball is in your court. We've had so many issues with those state machines, you'd need a much higher level explanations on the intent and effect of changing the state from PAUSED to PREPARED in *some* components. My main issue with your code is that the DAIs aren't touched and therefore it looks like a point solution, not a generic fix.

**Comment by zhigang-wu** [comment]
seems @tlauda refine this solution. I will close it now.

**Pull request closed without merging**